src/tail/session.rs

```diff
@@ -36,41 +48,96 @@ impl Session {
+                                if heartbeat_result.is_err() {
+                                    return heartbeat_result;
+                                }
+                                delay = delay_for(duration);
+                            }
+                        }
+                        _ => {
+                            return Ok(());
+                        }
                     }
-                    // This should loop forever until SIGINT is issued or Wrangler process is killed
-                    // through other means.
                 }
             }
             Err(e) => failure::bail!(http::format_error(e, None)),
         }
     }
 }
 
-async fn get_tunnel_url() -> Result<String, failure::Error> {
+async fn get_tunnel_url(rx: &mut Receiver<()>) -> Result<String, failure::Error> {
     // regex for extracting url from cloudflared metrics port.
     let url_regex = Regex::new("userHostname=\"(https://[a-z.-]+)\"").unwrap();
 
-    let mut attempt = 0;
+    struct RetryDelay {
+        delay: Delay,
+        attempt: u64,
+        max_attempts: u64,
+    }
 
-    // This exponential backoff retry loop retries retrieving the cloudflared endpoint url
-    // from the cloudflared /metrics endpoint until it gets the URL or has tried retrieving the URL
-    // over 5 times.
-    while attempt < 5 {
-        if let Ok(resp) = reqwest::get("http://localhost:8081/metrics").await {
-            let body = resp.text().await?;
+    // This retry loop retries retrieving the cloudflared endpoint url from the cloudflared /metrics
+    // until it gets the URL or has tried retrieving the URL over 5 times.
+    impl RetryDelay {
+        fn new(max_attempts: u64) -> Self {
+            RetryDelay {
+                delay: delay_for(Duration::from_millis(0)),
+                attempt: 0,
+                max_attempts,
+            }
+        }
 
-            for cap in url_regex.captures_iter(&body) {
-                return Ok(cap[1].to_string());
+        // our retry delay implements an [exponential backoff](TODO), which simply waits twice as
+        // long between each attempt.
+        fn reset(self) -> RetryDelay {
+            RetryDelay {
+                attempt: self.attempt + 1,
+                delay: delay_for(Duration::from_millis(self.attempt * self.attempt * 1000)),
+                max_attempts: self.max_attempts,
             }
         }
 
-        attempt += 1;
-        thread::sleep(Duration::from_millis(attempt * attempt * 1000));
+        fn expired(&self) -> bool {
+            self.attempt < self.max_attempts
+        }
+
+        fn is_elapsed(&self) -> bool {
+            self.delay.is_elapsed()
+        }
+    }
+
+    let mut delay = RetryDelay::new(5);
+
+    while !delay.expired() {
+        match rx.try_recv() {
+            Err(TryRecvError::Empty) => {
+                if delay.is_elapsed() {
+                    if let Ok(resp) = reqwest::get("http://localhost:8081/metrics").await {
```

Comment:
should make issue to make this configurable and also handle the error case (what happens if localhost:8081/metrics doesn't respond?

Reply:
there is already an issue for configurable ports. if there is no response it will keep retrying; really if any outcome happens that isn't the response being okay and containing the url, it will keep retrying until it runs out of retries.